In [1]:
#First lets install the missing packages

%pip install matplotlib
%pip install seaborn
%pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


In [2]:
# Import Libraries Required
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import numpy as np
import seaborn as sns
import sklearn

# Real State Transactions per Municipality in Belgium

We have been provided with a transactions dataset per municipality in Belgium (2010-2018), the dataset contains the Quarterly transactions per municipality but it also includes the aggregation by semester and year, at a high level only, basically it includes the number of transactions, the total amount and the total surface.

Our tasks:

1. Data Cleaning.

2. Data Preparation.


## Lets introduce CRISP-DM

CRISP-DM (Cross-Industry Standard Process for Data Mining) is a widely-used framework for structuring data mining and machine learning projects. It is designed to provide a systematic and organized approach to solving business problems using data-driven techniques.

The CRISP-DM process is divided into six main phases:

### Business Understanding:
In this phase, the team defines the problem or opportunity that the project will address, as well as the objectives and goals of the project.

### Data Understanding: 
In this phase, the team investigates the available data to determine what data is available and how it can be used to address the business problem.

### Data Preparation: 
In this phase, the team cleans and prepares the data for analysis. This can include tasks such as missing value imputation, outlier detection and removal, and data transformation.

### Modeling: 
In this phase, the team builds and tests machine learning models to find the best solution to the business problem.

### Evaluation: 
In this phase, the team assesses the performance of the chosen model and determines whether it meets the objectives and goals of the project.

### Deployment: 
In this phase, the team implements the chosen solution and makes it available for use in the business.

By following the CRISP-DM process, data scientists and business analysts can ensure that their projects are well-structured, well-documented, and aligned with the business objectives of the organization.

For the sake of this Portfolio Project only Data Understanding and Data Preparation phases will be done.

## Data Understanding

The second stage of the CRISP-DM process requires you to acquire the data listed in the project resources. This initial collection includes data loading, if this is necessary for data understanding. For example, if you use a specific tool for data understanding, it makes perfect sense to load your data into this tool. If you acquire multiple data sources then you need to consider how and when you're going to integrate these.


### Initial Data Report
Initial data collection report - List the data sources acquired together with their locations, the methods used to acquire them and any problems encountered. Record problems you encountered and any resolutions achieved. This will help both with future replication of this project and with the execution of similar future projects.



In [3]:
#Data source:  immo_by_municipality__2010-2019.txt
#Source Query location:  Azure Blob Storage

path =  'immo_by_municipality__2010-2019.txt'

# reads the data from the file - denotes as CSV, it has no header, sets column headers
df =  pd.read_csv(path, sep='|') 
df.head(10)

CD_YEAR         CD_TYPE_NL            CD_TYPE_FR  CD_REFNIS CD_REFNIS_NL  \
0     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
1     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
2     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
3     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
4     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
5     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
6     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
7     2010  gewone woonhuizen  maisons d'habitation      11002    ANTWERPEN   
8     2010  gewone woonhuizen  maisons d'habitation      11002    ANTWERPEN   
9     2010  gewone woonhuizen  maisons d'habitation      11002    ANTWERPEN   

  CD_REFNIS_FR CD_PERIOD CD_CLASS_SURFACE  MS_TOTAL_TRANSACTIONS  \
0   AARTSELAAR        Q1   totaal / total                     25   
1   AARTSELAAR        Q2   totaal / total                     21   
2   AARTSELAAR        Q3   totaal / total                     20   
3   AARTSELAAR        Q4   totaal / total                     24   
4   AARTSELAAR        S1   totaal / total                     46   
5   AARTSELAAR        S2   totaal / total                     44   
6   AARTSELAAR         Y   totaal / total                     90   
7       ANVERS        Q1   totaal / total                    746   
8       ANVERS        Q2   totaal / total                    797   
9       ANVERS        Q3   totaal / total                    865   

   MS_TOTAL_PRICE  MS_TOTAL_SURFACE  MS_MEAN_PRICE    MS_P10    MS_P25  \
0    6.207500e+06            7676.0   248300.00000  185000.0  225000.0   
1    5.548500e+06            6868.0   264214.28571  200000.0  212000.0   
2    5.317000e+06            5141.0   265850.00000  163750.0  232500.0   
3    6.004201e+06            9593.0   250175.05000  175000.0  210000.0   
4    1.175600e+07           14544.0   255565.21739  197500.0  220000.0   
5    1.132120e+07           14734.0   257300.02727  165000.0  220850.6   
6    2.307720e+07           29278.0   256413.34667  180000.0  220000.0   
7    1.590808e+08          125331.0   213245.10657  115000.0  150000.0   
8    1.695462e+08          141220.0   212730.54090  120000.0  153000.0   
9    1.910487e+08          161453.0   220865.50306  125000.0  157000.0   

     MS_P50    MS_P75    MS_P90  
0  247000.0  265000.0  320000.0  
1  252500.0  285000.0  300000.0  
2  252500.0  282500.0  370000.0  
3  245000.0  284500.0  334000.0  
4  251000.0  270000.0  320000.0  
5  250000.0  282500.0  340000.0  
6  250000.0  280000.0  332000.0  
7  192500.0  250000.0  320000.0  
8  193000.0  247000.0  329000.0  
9  205000.0  264000.0  335000.0

### 2.2 Describe Data
Data description report - Describe the data that has been acquired including its format, its quantity (for example, the number of records and fields in each table), the identities of the fields and any other surface features which have been discovered. Evaluate whether the data acquired satisfies your requirements.

In [4]:
#List the columns
df.columns

Index(['CD_YEAR', 'CD_TYPE_NL', 'CD_TYPE_FR', 'CD_REFNIS', 'CD_REFNIS_NL',
       'CD_REFNIS_FR', 'CD_PERIOD', 'CD_CLASS_SURFACE',
       'MS_TOTAL_TRANSACTIONS', 'MS_TOTAL_PRICE', 'MS_TOTAL_SURFACE',
       'MS_MEAN_PRICE', 'MS_P10', 'MS_P25', 'MS_P50', 'MS_P75', 'MS_P90'],
      dtype='object')

In [5]:
#Lets review the number of rows and columns
df.shape

(112160, 17)

In [6]:
#And the data types
df.dtypes

CD_YEAR                    int64
CD_TYPE_NL                object
CD_TYPE_FR                object
CD_REFNIS                  int64
CD_REFNIS_NL              object
CD_REFNIS_FR              object
CD_PERIOD                 object
CD_CLASS_SURFACE          object
MS_TOTAL_TRANSACTIONS      int64
MS_TOTAL_PRICE           float64
MS_TOTAL_SURFACE         float64
MS_MEAN_PRICE            float64
MS_P10                   float64
MS_P25                   float64
MS_P50                   float64
MS_P75                   float64
MS_P90                   float64
dtype: object

In [7]:
#Let Pandas calculate some basic statistics
df.describe()

CD_YEAR      CD_REFNIS  MS_TOTAL_TRANSACTIONS  MS_TOTAL_PRICE  \
count  112160.000000  112160.000000          112160.000000    6.329800e+04   
mean     2013.296719   46436.990594              29.383425    1.046981e+07   
std         2.291938   24120.236728              93.230464    2.564732e+07   
min      2010.000000   11001.000000               1.000000    1.280000e+05   
25%      2011.000000   24066.000000               5.000000    2.893500e+06   
50%      2013.000000   44049.000000              12.000000    4.995500e+06   
75%      2015.000000   63058.000000              27.000000    9.751412e+06   
max      2018.000000   93090.000000            5699.000000    1.073798e+09   

       MS_TOTAL_SURFACE  MS_MEAN_PRICE         MS_P10        MS_P25  \
count      50297.000000   6.329800e+04   63298.000000  6.329800e+04   
mean       31985.658325   2.006188e+05  118794.353623  1.493028e+05   
std        33419.918658   1.248665e+05   75927.311385  9.022346e+04   
min          139.000000   1.393958e+01       7.281553  9.756098e+00   
25%        13223.000000   1.480562e+05   75000.000000  1.037094e+05   
50%        21962.000000   1.982737e+05  117000.000000  1.500000e+05   
75%        38546.000000   2.571778e+05  160000.000000  1.975000e+05   
max       755681.000000   1.580633e+06  820000.000000  1.005000e+06   

             MS_P50        MS_P75        MS_P90  
count  6.329800e+04  6.329800e+04  6.329800e+04  
mean   1.881678e+05  2.367883e+05  2.982554e+05  
std    1.122684e+05  1.494275e+05  2.089012e+05  
min    1.208897e+01  1.795143e+01  1.900000e+01  
25%    1.400000e+05  1.763340e+05  2.150000e+05  
50%    1.875000e+05  2.350000e+05  2.850000e+05  
75%    2.450000e+05  3.000000e+05  3.750000e+05  
max    1.337500e+06  2.146760e+06  4.391000e+06

In [8]:
# And lets review how many rows are not null for each column together with data types
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112160 entries, 0 to 112159
Data columns (total 17 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   CD_YEAR                112160 non-null  int64  
 1   CD_TYPE_NL             112160 non-null  object 
 2   CD_TYPE_FR             112160 non-null  object 
 3   CD_REFNIS              112160 non-null  int64  
 4   CD_REFNIS_NL           112160 non-null  object 
 5   CD_REFNIS_FR           112160 non-null  object 
 6   CD_PERIOD              112160 non-null  object 
 7   CD_CLASS_SURFACE       112160 non-null  object 
 8   MS_TOTAL_TRANSACTIONS  112160 non-null  int64  
 9   MS_TOTAL_PRICE         63298 non-null   float64
 10  MS_TOTAL_SURFACE       50297 non-null   float64
 11  MS_MEAN_PRICE          63298 non-null   float64
 12  MS_P10                 63298 non-null   float64
 13  MS_P25                 63298 non-null   float64
 14  MS_P50                 63298 non-nul

In [9]:
# Finally lets check the first 5 rows to see how the dataframe looks lie
df.head(5)

CD_YEAR         CD_TYPE_NL            CD_TYPE_FR  CD_REFNIS CD_REFNIS_NL  \
0     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
1     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
2     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
3     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   
4     2010  gewone woonhuizen  maisons d'habitation      11001   AARTSELAAR   

  CD_REFNIS_FR CD_PERIOD CD_CLASS_SURFACE  MS_TOTAL_TRANSACTIONS  \
0   AARTSELAAR        Q1   totaal / total                     25   
1   AARTSELAAR        Q2   totaal / total                     21   
2   AARTSELAAR        Q3   totaal / total                     20   
3   AARTSELAAR        Q4   totaal / total                     24   
4   AARTSELAAR        S1   totaal / total                     46   

   MS_TOTAL_PRICE  MS_TOTAL_SURFACE  MS_MEAN_PRICE    MS_P10    MS_P25  \
0       6207500.0            7676.0   248300.00000  185000.0  225000.0   
1       5548500.0            6868.0   264214.28571  200000.0  212000.0   
2       5317000.0            5141.0   265850.00000  163750.0  232500.0   
3       6004201.2            9593.0   250175.05000  175000.0  210000.0   
4      11756000.0           14544.0   255565.21739  197500.0  220000.0   

     MS_P50    MS_P75    MS_P90  
0  247000.0  265000.0  320000.0  
1  252500.0  285000.0  300000.0  
2  252500.0  282500.0  370000.0  
3  245000.0  284500.0  334000.0  
4  251000.0  270000.0  320000.0

In [10]:
unique_types = df["CD_TYPE_FR"].unique()
print(unique_types)

# We are only interested in houses and apparment Types, villas or terrains are not relevant for us

["maisons d'habitation" 'villas, bungalows, maisons de campagne'
 'appartements, flats, studios' 'terrains � b�tir']


## Data Preparation

Data Preparation & Cleaning is an extremely important
part of the overall Machine Learning process, one that must
be considered before ever looking to build or train a model.
A common phrase in Machine Learning is...
"Garbage in...Garbage Out!"

If the data isn't clean or isn’t prepared in an appropriate way,
even the fanciest of algorithms or models will struggle to
learn.
On top of this, ensuring the data is clean can actually be one
of the biggest boosters of model performance and accuracy!

### 8-Step Checklist

Thanks to Data Science Infinity Ebook here: https://cdn.fs.teachablecdn.com/KTJythseReJyqGoIjyQg

1. Missing values.
2. Duplicate & Low Variation Data.
3. Incorrect & Irrelevant Data.
4. Categorical Data.
5. Outliers. 
6. Feature Scaling.
7. Feature Engineering/Selection.
8. Validation Split.

Note: Which steps are applicable can depend on the data you’re
using, the problem you’re solving, and on the type of model you’re
applying! However, in the vast majority of cases you can’t go
wrong at least considering these 8 steps!


#### Missing Values

In many cases, your chosen model or algorithm simply
won't know how to process missing values - and you will
be returned an error.

Even if the missing values don't lead to an error you do
always want to ensure that you are passing the model
the most useful information to learn from - and you
should consider whether missing values meet that criteria
or not.

The two most common approaches for dealing with
missing values:

**Removal:** Often you will simply remove any observations
(rows) where one or missing values are present. You can
also remove entire columns if no information is present.

**Imputation:** This is where you input or "impute"
replacement values where they were originally missing.
This can be based upon the column mean, median, or
most common value or more advanced approaches that
take into account other present data points to give an
estimation of what the missing value might be!


In [11]:
#Lets see how many observations have missing values per feature
df.isnull().sum()

CD_YEAR                      0
CD_TYPE_NL                   0
CD_TYPE_FR                   0
CD_REFNIS                    0
CD_REFNIS_NL                 0
CD_REFNIS_FR                 0
CD_PERIOD                    0
CD_CLASS_SURFACE             0
MS_TOTAL_TRANSACTIONS        0
MS_TOTAL_PRICE           48862
MS_TOTAL_SURFACE         61863
MS_MEAN_PRICE            48862
MS_P10                   48862
MS_P25                   48862
MS_P50                   48862
MS_P75                   48862
MS_P90                   48862
dtype: int64

In [12]:
#Lets define a reusable function to list the missing values per field and its percentage over all data points.
def missing_values_table(df):
        mis_val = df.isnull().sum()
        mis_val_percent = 100 * df.isnull().sum() / len(df)
        mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
        mis_val_table_ren_columns = mis_val_table.rename(
        columns = {0 : 'Missing Values', 1 : '% of Total Values'})
        mis_val_table_ren_columns = mis_val_table_ren_columns[
            mis_val_table_ren_columns.iloc[:,1] != 0].sort_values(
        '% of Total Values', ascending=False).round(1)
        print ("Your selected dataframe has " + str(df.shape[1]) + " columns.\n"      
            "There are " + str(mis_val_table_ren_columns.shape[0]) +
              " columns that have missing values.")
        return mis_val_table_ren_columns

In [13]:
# Lets execute that function
missing_values_table(df)

Your selected dataframe has 17 columns.
There are 8 columns that have missing values.


Missing Values  % of Total Values
MS_TOTAL_SURFACE           61863               55.2
MS_TOTAL_PRICE             48862               43.6
MS_MEAN_PRICE              48862               43.6
MS_P10                     48862               43.6
MS_P25                     48862               43.6
MS_P50                     48862               43.6
MS_P75                     48862               43.6
MS_P90                     48862               43.6

In [14]:
# Get the columns with > 50% missing
# We could for example remove entire columns depending on the percentage of missing values, but here the code is commented for your information
# As another approach will be used, keep reading

# missing_df = missing_values_table(df);
# missing_columns = list(missing_df[missing_df['% of Total Values'] > 50].index)
# print('We will remove %d columns.' % len(missing_columns))

We will drop the percentile columns, MS_P10, MS_P25, MS_P50, MS_P75, MS_P90 as they are irrelevant for our analysis. We also remove text fields which are not needed, more specifically those ones with NL suffix which are in dutch, we will leave the city name and property type features in french.

We have the total price the total surface, which will allow us to get the price per square meter.  However the total surface is null  on  55.2% of the rows.  By manual analysis we can see that apparently this is empty on rows where the period is Q1-Q4 or S1-S2. As we want to do a yearly analysis and not per quarter or per semester, lets drop these rows  to see if we have a cleaner dataset on the yearly basis.

In [15]:
#Drop unnecesary fields as explained above
df = df.drop(columns=['MS_P10', 'MS_P25', 'MS_P50', 'MS_P75', 'MS_P90','CD_TYPE_NL','CD_REFNIS','CD_REFNIS_NL'])

#Drop unnecesary rows as explained above, we are only interested on yearly data points.
df = df.drop(df[df['CD_PERIOD'] != 'Y'].index)


In [16]:
#And then check again if we have null values.
df.isnull().sum()

CD_YEAR                     0
CD_TYPE_FR                  0
CD_REFNIS_FR                0
CD_PERIOD                   0
CD_CLASS_SURFACE            0
MS_TOTAL_TRANSACTIONS       0
MS_TOTAL_PRICE           2798
MS_TOTAL_SURFACE         5669
MS_MEAN_PRICE            2798
dtype: int64

Without a total price and total surface we wont be able to calculate a square meter average price. However when analyzing the dataset manually, I found out that total surface feature  is only available for type House data points (maissons), for appartments this field is always empty. So lets split the dataframe in 2 for easier analysis.



In [17]:
#Split the dataframe in one for houses and another for appartements
df_houses = df[df['CD_TYPE_FR'] == "maisons d'habitation"]
#Lets save the dataset as a new file for easier loading and processing later.
df_houses.to_csv('df_houses.csv', index=False)

df_appartements = df[df['CD_TYPE_FR'] == 'appartements, flats, studios'] 
#Lets save the dataset as a new file for easier loading and processing later.
df_appartements.to_csv('df_houses.csv', index=False)

In [18]:
missing_values_table(df_houses)

Your selected dataframe has 9 columns.
There are 3 columns that have missing values.


Missing Values  % of Total Values
MS_TOTAL_PRICE                61                1.3
MS_TOTAL_SURFACE              61                1.3
MS_MEAN_PRICE                 61                1.3

In [19]:
missing_values_table(df_appartements)

Your selected dataframe has 9 columns.
There are 3 columns that have missing values.


Missing Values  % of Total Values
MS_TOTAL_SURFACE            4271              100.0
MS_TOTAL_PRICE              1400               32.8
MS_MEAN_PRICE               1400               32.8

In [20]:
#For the houses dataframe Lets drop the rows where both total price and total surface are null at the same time
df_houses = df_houses.dropna(subset=['MS_TOTAL_PRICE', 'MS_TOTAL_SURFACE'], how='any')

#And now lets check the percentage of missing values per feature.
missing_values_table(df_houses)


Your selected dataframe has 9 columns.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Missing Values, % of Total Values]
Index: []

In [21]:
#For the appartments dataframe Lets drop first the column total surface as all values are empty
df_appartements = df_appartements.drop(columns=['MS_TOTAL_SURFACE'])

#For the appartments dataframe also Lets drop the rows where both total price and mean price are empty
df_appartements = df_appartements.dropna(subset=['MS_TOTAL_PRICE', 'MS_MEAN_PRICE'], how='any')
missing_values_table(df_appartements)


Your selected dataframe has 8 columns.
There are 0 columns that have missing values.


Empty DataFrame
Columns: [Missing Values, % of Total Values]
Index: []

Lets describe again our dataframes

In [22]:
df_houses.columns


Index(['CD_YEAR', 'CD_TYPE_FR', 'CD_REFNIS_FR', 'CD_PERIOD',
       'CD_CLASS_SURFACE', 'MS_TOTAL_TRANSACTIONS', 'MS_TOTAL_PRICE',
       'MS_TOTAL_SURFACE', 'MS_MEAN_PRICE'],
      dtype='object')

In [23]:
df_appartements.columns

Index(['CD_YEAR', 'CD_TYPE_FR', 'CD_REFNIS_FR', 'CD_PERIOD',
       'CD_CLASS_SURFACE', 'MS_TOTAL_TRANSACTIONS', 'MS_TOTAL_PRICE',
       'MS_MEAN_PRICE'],
      dtype='object')

In [24]:
df_houses.shape

(4644, 9)

In [25]:
df_appartements.shape

(2871, 8)

In [26]:
df_houses.dtypes

CD_YEAR                    int64
CD_TYPE_FR                object
CD_REFNIS_FR              object
CD_PERIOD                 object
CD_CLASS_SURFACE          object
MS_TOTAL_TRANSACTIONS      int64
MS_TOTAL_PRICE           float64
MS_TOTAL_SURFACE         float64
MS_MEAN_PRICE            float64
dtype: object

In [27]:
df_appartements.dtypes

CD_YEAR                    int64
CD_TYPE_FR                object
CD_REFNIS_FR              object
CD_PERIOD                 object
CD_CLASS_SURFACE          object
MS_TOTAL_TRANSACTIONS      int64
MS_TOTAL_PRICE           float64
MS_MEAN_PRICE            float64
dtype: object

In [28]:
df_houses.describe()

CD_YEAR  MS_TOTAL_TRANSACTIONS  MS_TOTAL_PRICE  MS_TOTAL_SURFACE  \
count  4644.000000            4644.000000    4.644000e+03       4644.000000   
mean   2013.500000             110.769165    2.206725e+07      61856.257289   
std       2.292662             205.485428    4.452167e+07      57171.806895   
min    2010.000000              10.000000    1.217500e+06       3206.000000   
25%    2012.000000              36.000000    6.637819e+06      29942.750000   
50%    2013.500000              62.000000    1.196470e+07      47386.000000   
75%    2015.000000             119.000000    2.311122e+07      75668.250000   
max    2017.000000            3539.000000    8.275591e+08     755681.000000   

       MS_MEAN_PRICE  
count    4644.000000  
mean   197431.718164  
std     61213.966342  
min     85874.026749  
25%    156607.604650  
50%    188701.402215  
75%    225734.454175  
max    690386.601090

In [29]:
df_appartements.describe()

CD_YEAR  MS_TOTAL_TRANSACTIONS  MS_TOTAL_PRICE  MS_MEAN_PRICE
count  2871.000000            2871.000000    2.871000e+03    2871.000000
mean   2013.412052             115.660745    2.419658e+07  200165.388617
std       2.277258             323.841204    6.649911e+07   49274.600797
min    2010.000000              10.000000    8.530000e+05   81136.363636
25%    2011.000000              19.000000    3.627898e+06  169479.803920
50%    2013.000000              38.000000    7.523552e+06  195206.496640
75%    2015.000000              90.000000    1.833507e+07  223877.693840
max    2017.000000            5699.000000    1.073798e+09  543859.933210

In [30]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 16529 entries, 6 to 110570
Data columns (total 9 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   CD_YEAR                16529 non-null  int64  
 1   CD_TYPE_FR             16529 non-null  object 
 2   CD_REFNIS_FR           16529 non-null  object 
 3   CD_PERIOD              16529 non-null  object 
 4   CD_CLASS_SURFACE       16529 non-null  object 
 5   MS_TOTAL_TRANSACTIONS  16529 non-null  int64  
 6   MS_TOTAL_PRICE         13731 non-null  float64
 7   MS_TOTAL_SURFACE       10860 non-null  float64
 8   MS_MEAN_PRICE          13731 non-null  float64
dtypes: float64(3), int64(2), object(4)
memory usage: 1.3+ MB


In [31]:
df.head(500)

CD_YEAR            CD_TYPE_FR CD_REFNIS_FR CD_PERIOD CD_CLASS_SURFACE  \
6        2010  maisons d'habitation   AARTSELAAR         Y   totaal / total   
13       2010  maisons d'habitation       ANVERS         Y   totaal / total   
20       2010  maisons d'habitation     BOECHOUT         Y   totaal / total   
27       2010  maisons d'habitation         BOOM         Y   totaal / total   
34       2010  maisons d'habitation     BORSBEEK         Y   totaal / total   
...       ...                   ...          ...       ...              ...   
3469     2010  maisons d'habitation       BILZEN         Y   totaal / total   
3476     2010  maisons d'habitation         LOOZ         Y   totaal / total   
3483     2010  maisons d'habitation        HEERS         Y   totaal / total   
3486     2010  maisons d'habitation    HERSTAPPE         Y   totaal / total   
3493     2010  maisons d'habitation      HOESELT         Y   totaal / total   

      MS_TOTAL_TRANSACTIONS  MS_TOTAL_PRICE  MS_TOTAL_SURFACE  MS_MEAN_PRICE  
6                        90    2.307720e+07           29278.0   256413.34667  
13                     3179    6.842183e+08          597584.0   215230.68177  
20                       68    1.986547e+07          113712.0   292139.26471  
27                      168    2.833533e+07           28676.0   168662.67857  
34                       60    1.430926e+07           19680.0   238487.74783  
...                     ...             ...               ...            ...  
3469                    116    2.006851e+07          116844.0   173004.36638  
3476                     78    1.269005e+07           72003.0   162692.94872  
3483                     56    8.288100e+06           49266.0   148001.78571  
3486                      1             NaN               NaN            NaN  
3493                     44    7.927599e+06           67687.0   180172.70114  

[500 rows x 9 columns]

#### Duplicate & Low Variation Data

When looking through your data, don't just look for
missing values - keep an eye out for duplicate data, or
data that has low variation.
Duplicate data is most commonly rows of data that are
exactly the same across all columns. This duplicate rows
do not add anything to the learning process of the model
or algorithm, but do add storage & processing overhead.
In the vast majority of cases you can remove duplicate
rows prior to training your model.
Low variation data is where a column in your dataset
contain only one (or few) unique value(s).
An example: There is a column in a house price dataset
called "property_type". Every row in this column has the
value "house". This column won’t add any value to the
learning process, so can be removed.

In our dataset CD_CLASS_SURFACE has always the same value through the entire dataset, so we remove the entire feature.


In [32]:
df_houses = df_houses.drop(columns=['CD_CLASS_SURFACE'])
df_appartements = df_appartements.drop(columns=['CD_CLASS_SURFACE'])


#### Incorrect & Irrelevant Data

Irrelevant data is anything that isn’t related specifically
to the problem you’re looking to solve. For example, if
you're predicting house prices, but your dataset contains
commercial properties as well - these would need to be
removed.
Incorrect data can be hard to spot! An example could
be looking for values that shouldn’t be possible such a
negative house price values.
For categorical or text variables you should spend time
analysing the unique values within a column.
For example, you were predicting car prices and had a
column in the dataset called "car_colour". Upon
inspection you find values of "Orange", "orange", "orang"
are all present. These need to be rectifed prior to training
otherwise you will be limited the potential learning that an
take place!
Always explore your data thoroughly!

Our CD_TYPE_FR has the following unique values:


In [33]:
unique_types = df["CD_TYPE_FR"].unique()
print(unique_types)

["maisons d'habitation" 'villas, bungalows, maisons de campagne'
 'appartements, flats, studios' 'terrains � b�tir']


For our analysis, we are interested only in houses, not terrains or villas, etc. During our missing values phase we split the original dataframe in 2, one for houses and other for appartments, so the incorrect & irrelevant data step is already done.


#### Categorical Data

Generally speaking, ML models like being fed numerical
data. They are less fond of categorical data!
Categorical data is anything that is listed as groups,
classes, or text. A simple example would be a column for
gender which contains values of either "Male" or "Female".
Your model or algorithm won’t know how to assign some
numerical importance to these values, so you often want
to turn these groups or classes into numerical values.
A common approach is called One Hot Encoding where
you create new columns, one for each unique class in your
categorical column. You fill these new columns with
values of 1 or 0 depending on which is true for each
observation.
Other encoding techniques you can consider are; Label
Encoding, Binary Encoding, Target Encoding, Ordinal
Encoding, & Feature Hashing.

The only categoricla column is the municipality name, as there are over 500 municipalities in Belgium, we will use Label encoding in this case.

In label encoding in Python, we replace the categorical value with a numeric value between 0 and the number of classes minus 1. If the categorical variable value contains 5 distinct classes, we use (0, 1, 2, 3, and 4). We will keep for now the string field and encoded field.


In [34]:

# Import label encoder
from sklearn import preprocessing
# # label_encoder object knows how to understand word labels.
label_encoder = preprocessing.LabelEncoder()
  
# Encode labels in column 'species'.
df_appartements['CD_REFNIS_FR_ENCODED']= label_encoder.fit_transform(df_appartements['CD_REFNIS_FR'])
df_houses['CD_REFNIS_FR_ENCODED']= label_encoder.fit_transform(df_houses['CD_REFNIS_FR'])
   
df_appartements['CD_REFNIS_FR_ENCODED'].unique()
df_houses['CD_REFNIS_FR_ENCODED'].unique()

array([  2,  17,  68,  70,  73,  83,  84, 142, 153, 225, 253, 270, 273,
       284, 329, 383, 394, 440, 458, 471, 472, 473, 491, 559, 564, 566,
       569, 576, 586, 349,  51,  69,  72, 139, 223, 324, 348, 397, 433,
       434, 477, 479, 560,  20,  36,  37,  44, 126, 187, 204, 230, 231,
       234, 246, 255, 275, 325, 356, 365, 373, 401, 414, 441, 446, 449,
       519, 532, 533, 553, 299,  11,  31,  49,  90, 157, 159, 175, 184,
       261, 266, 282, 374, 461, 463, 470, 520, 546, 562, 563,  23,  45,
        61, 130, 182, 198, 203, 208, 232, 241, 271, 272, 320, 336, 346,
       358, 363, 408, 418, 423, 483, 494, 501, 527, 570, 574, 452, 138,
       289, 328, 447, 551, 556, 312,   3,   1,  46,  47,  55,  60,  71,
        77, 129, 189, 205, 229, 240, 243, 254, 276, 286, 287, 302, 339,
       341, 413, 453, 502, 511, 515, 306, 330, 379, 508, 197,  42,  78,
        79, 103, 113, 191, 202, 258, 260, 267, 295, 378, 399, 426, 450,
       518, 525, 545, 547, 100, 224, 304, 410, 412, 439, 442, 54

#### Outliers

There is no formal definition for an outlier. You can think
of them as any data point that is very different to the
majority.
How you deal with outliers is dependent on the problem
you are solving, and the model you are applying. For
example, if your data contained one value that was 1000x
any other, this could badly affect a Linear Regression
model which tries to generalise a rule across all
observations. A Decision Tree would be unaffected
however, as it deals with each observation independently.
In practice, outliers are commonly isolated using the
number of standard deviations from the mean, or a rule
based upon the interquartile range.
In cases where you want to mitigate the effects of outliers,
you may look to simply remove any observations (rows)
that contain outlier values in one or more of the columns
or you may look to replace their values to reduce their
effect.
Always remember - just because a value is very high, or
very low, that does not mean it is wrong to be included.

In [35]:
# We wont remove outliers because our dataset is already aggregated at the yearly level,
# we dont have each real state transaction to be able to see Outliers.

#### Feature Scaling

Feature Scaling is where you force all the values from a
column in your data to exist on the same scale. In certain
scenarios it will help the model assess the relationships
between variables more fairly, and more accurately.
The two most common scaling techniques are:
Standardisation: rescales all values to have a mean of 0
and standard deviation of 1. In other words, the majority
of your values end up between -4 and +4
Normalisation: rescales data so that it exists in a range
between 0 and 1
Feature Scaling is essential for distance-based models
such as k-means or k-nearest-neighbours.
Feature Scaling is recommended for any algorithms that
utilise Gradient Descent such as Linear Regression,
Logistic Regression, and Neural Networks.
Feature Scaling is not necessary for tree-based
algorithms such as Decision Trees & Random Forests.

To scale the values in  a Pandas DataFrame, you can use the StandardScaler class from scikit-learn library. 

This will scale the values in the specified columns to have a mean of 0 and a standard deviation of 1. You can also use the MinMaxScaler class to scale the values to a specific range (e.g. 0 to 1).

It's important to note that you should only scale the values in a dataset if it is necessary for the analysis or machine learning task you are performing. Scaling the values can be useful if the features in the dataset have different units or scales, as it can help to put them on a common scale and prevent features with larger scales from dominating the analysis. However, scaling the values can also change the interpretation of the features, so you should carefully consider whether it is appropriate for your specific case.

The scaled features are saved in a copy of the dataframe to keep the original values, the reason why, its at this point we dont know if the Time Series algorightms we will use require the numeric features to be scaled or not.

In [36]:
from sklearn.preprocessing import StandardScaler

# Select the columns you want to scale
columns_to_scale = ['MS_TOTAL_TRANSACTIONS', 'MS_TOTAL_PRICE','MS_TOTAL_SURFACE','MS_MEAN_PRICE']

# Create a scaler object
scaler = StandardScaler()

# Fit the scaler to the data
scaler.fit(df_houses[columns_to_scale])

df_houses_scaled = df_houses.copy()

# Transform the data
df_houses_scaled[columns_to_scale] = scaler.transform(df_houses_scaled[columns_to_scale])
df_houses_scaled
#df_appartements_scaled 

CD_YEAR            CD_TYPE_FR   CD_REFNIS_FR CD_PERIOD  \
6          2010  maisons d'habitation     AARTSELAAR         Y   
13         2010  maisons d'habitation         ANVERS         Y   
20         2010  maisons d'habitation       BOECHOUT         Y   
27         2010  maisons d'habitation           BOOM         Y   
34         2010  maisons d'habitation       BORSBEEK         Y   
...         ...                   ...            ...       ...   
103202     2017  maisons d'habitation        DOISCHE         Y   
103209     2017  maisons d'habitation      FLORENNES         Y   
103216     2017  maisons d'habitation  PHILIPPEVILLE         Y   
103223     2017  maisons d'habitation       WALCOURT         Y   
103230     2017  maisons d'habitation      VIROINVAL         Y   

        MS_TOTAL_TRANSACTIONS  MS_TOTAL_PRICE  MS_TOTAL_SURFACE  \
6                   -0.101085        0.022687         -0.569892   
13                  14.933230       14.874160          9.371497   
20                  -0.208160       -0.049459          0.907114   
27                   0.278545        0.140802         -0.580423   
34                  -0.247096       -0.174271         -0.737790   
...                       ...             ...               ...   
103202              -0.432044       -0.438145         -0.837133   
103209              -0.203293       -0.283145         -0.285648   
103216              -0.334703       -0.344021         -0.392705   
103223              -0.110819       -0.197193          0.347127   
103230              -0.276298       -0.382199         -0.532037   

        MS_MEAN_PRICE  CD_REFNIS_FR_ENCODED  
6            0.963636                     2  
13           0.290798                    17  
20           1.547322                    68  
27          -0.470026                    70  
34           0.670769                    73  
...               ...                   ...  
103202      -1.322700                   135  
103209      -0.985078                   172  
103216      -0.598913                   427  
103223      -0.758441                   539  
103230      -1.696839                   528  

[4644 rows x 9 columns]

In [37]:
from sklearn.preprocessing import StandardScaler

# Select the columns you want to scale
columns_to_scale = ['MS_TOTAL_TRANSACTIONS', 'MS_TOTAL_PRICE','MS_MEAN_PRICE']

# Create a scaler object
scaler = StandardScaler()

# Fit the scaler to the data
scaler.fit(df_appartements[columns_to_scale])

df_appartements_scaled = df_appartements.copy()

# Transform the data
df_appartements_scaled[columns_to_scale] = scaler.transform(df_appartements_scaled[columns_to_scale])
df_appartements_scaled


CD_YEAR                    CD_TYPE_FR   CD_REFNIS_FR CD_PERIOD  \
8004       2010  appartements, flats, studios     AARTSELAAR         Y   
8011       2010  appartements, flats, studios         ANVERS         Y   
8018       2010  appartements, flats, studios       BOECHOUT         Y   
8025       2010  appartements, flats, studios           BOOM         Y   
8032       2010  appartements, flats, studios       BORSBEEK         Y   
...         ...                           ...            ...       ...   
110458     2017  appartements, flats, studios        EGHEZEE         Y   
110486     2017  appartements, flats, studios          NAMUR         Y   
110498     2017  appartements, flats, studios  PROFONDEVILLE         Y   
110512     2017  appartements, flats, studios    SAMBREVILLE         Y   
110537     2017  appartements, flats, studios       GEMBLOUX         Y   

        MS_TOTAL_TRANSACTIONS  MS_TOTAL_PRICE  MS_MEAN_PRICE  \
8004                -0.097783       -0.093766       0.277511   
8011                14.155511       10.554175      -0.927251   
8018                -0.181172       -0.183208       0.215833   
8025                -0.057633       -0.086058      -0.196966   
8032                -0.144110       -0.208224      -1.017593   
...                       ...             ...            ...   
110458              -0.323242       -0.329526       0.157550   
110486               0.649627        0.565073      -0.217114   
110498              -0.301622       -0.320685      -0.820914   
110512              -0.267649       -0.302449      -1.201987   
110537              -0.184260       -0.204043      -0.209875   

        CD_REFNIS_FR_ENCODED  
8004                       2  
8011                      11  
8018                      53  
8025                      55  
8032                      58  
...                      ...  
110458                   108  
110486                   284  
110498                   320  
110512                   349  
110537                   131  

[2871 rows x 8 columns]

#### Feature Engineering & Selection

Feature Engineering is the process of using further
knowledge to supplement or transform the original
feature set.
The key to good Feature Engineering is to create or refine
features that the algorithm or model can understand
better or that it will find more useful than the raw features
for solving the particular problem at hand.
Feature Selection is where you only keep a subset of the
most informative variables. This can be done using
human intuition, or dynamically based upon statistical
analysis.
A smaller feature set can lead to improved model
accuracy through reduced noise. It can mean a lower
computational cost, and improved processing speed. It
can also make your models easier to understand &
explain to stakeholders & customers!

For this problem lets add the average price per square meter to the houses dataframe

In [38]:
def price_per_surface(row):
  return row['MS_TOTAL_PRICE'] / row['MS_TOTAL_SURFACE']

# Add the new column to the dataframe
df_houses = df_houses.assign(price_per_surface=df.apply(price_per_surface, axis=1))
df_houses

CD_YEAR            CD_TYPE_FR   CD_REFNIS_FR CD_PERIOD  \
6          2010  maisons d'habitation     AARTSELAAR         Y   
13         2010  maisons d'habitation         ANVERS         Y   
20         2010  maisons d'habitation       BOECHOUT         Y   
27         2010  maisons d'habitation           BOOM         Y   
34         2010  maisons d'habitation       BORSBEEK         Y   
...         ...                   ...            ...       ...   
103202     2017  maisons d'habitation        DOISCHE         Y   
103209     2017  maisons d'habitation      FLORENNES         Y   
103216     2017  maisons d'habitation  PHILIPPEVILLE         Y   
103223     2017  maisons d'habitation       WALCOURT         Y   
103230     2017  maisons d'habitation      VIROINVAL         Y   

        MS_TOTAL_TRANSACTIONS  MS_TOTAL_PRICE  MS_TOTAL_SURFACE  \
6                          90    2.307720e+07          29278.00   
13                       3179    6.842183e+08         597584.00   
20                         68    1.986547e+07         113712.00   
27                        168    2.833533e+07          28676.00   
34                         60    1.430926e+07          19680.00   
...                       ...             ...               ...   
103202                     22    2.562400e+06          14001.00   
103209                     69    9.462500e+06          45527.00   
103216                     42    6.752500e+06          39407.04   
103223                     88    1.328884e+07          81700.00   
103230                     54    5.052925e+06          31442.00   

        MS_MEAN_PRICE  CD_REFNIS_FR_ENCODED  price_per_surface  
6       256413.346670                     2         788.209618  
13      215230.681770                    17        1144.974326  
20      292139.264710                    68         174.699856  
27      168662.678570                    70         988.120031  
34      238487.747830                    73         727.096792  
...               ...                   ...                ...  
103202  116472.727270                   135         183.015499  
103209  137137.681160                   172         207.843697  
103216  160773.809520                   427         171.352631  
103223  151009.545450                   539         162.654100  
103230   93572.685185                   528         160.706221  

[4644 rows x 10 columns]

What about if we want to see the % increase of transactions per year per city, and do the same for price per square meter.

In this case we need to take year 2010 as the basis, and then do the calculation per row per year per city.

In [47]:
# Group the data by city and year
df_grouped_houses_totalTransactions = df_houses.groupby(['CD_REFNIS_FR', 'CD_YEAR'])['MS_TOTAL_TRANSACTIONS'].sum().reset_index()
df_grouped_houses_price_per_surface = df_houses.groupby(['CD_REFNIS_FR', 'CD_YEAR'])['price_per_surface'].sum().reset_index()

# Calculate the percentage change in transactions per year for each city
df_grouped_houses_totalTransactions['pct_change_transactions'] = df_grouped_houses_totalTransactions.groupby('CD_REFNIS_FR')['MS_TOTAL_TRANSACTIONS'].pct_change()*100
df_grouped_houses_price_per_surface['pct_change_pricepersqm'] = df_grouped_houses_price_per_surface.groupby('CD_REFNIS_FR')['price_per_surface'].pct_change()*100

# Filter by City to see the % change per year
#df_grouped_houses_anvers = df_grouped_houses.loc[df_grouped_houses['CD_REFNIS_FR'] == 'ANVERS']
#df_grouped_houses = df_grouped_houses.sort_values('CD_YEAR', ascending=True)
#df_grouped_houses_anvers

In [52]:
pd.set_option('display.max_rows', 100)
df_grouped_houses_price_per_surface = df_grouped_houses_price_per_surface.fillna(0)

df_grouped_houses_price_per_surface.head(100)

CD_REFNIS_FR  CD_YEAR  price_per_surface  pct_change_pricepersqm
0           AALTER     2010         292.528065                0.000000
1           AALTER     2011         238.854558              -18.348157
2           AALTER     2012         147.676510              -38.173041
3           AALTER     2013         193.593618               31.093035
4           AALTER     2014         248.690774               28.460213
5           AALTER     2015         327.467359               31.676521
6           AALTER     2016         305.742010               -6.634356
7           AALTER     2017         468.156627               53.121459
8         AARSCHOT     2010         173.703680                0.000000
9         AARSCHOT     2011         201.074901               15.757422
10        AARSCHOT     2012         176.486099              -12.228678
11        AARSCHOT     2013         193.847443                9.837230
12        AARSCHOT     2014         202.116502                4.265756
13        AARSCHOT     2015         183.997991               -8.964390
14        AARSCHOT     2016         236.458317               28.511358
15        AARSCHOT     2017         276.695810               17.016738
16      AARTSELAAR     2010         788.209618                0.000000
17      AARTSELAAR     2011         878.700706               11.480587
18      AARTSELAAR     2012         623.582440              -29.033579
19      AARTSELAAR     2013         849.785350               36.274740
20      AARTSELAAR     2014         772.641082               -9.078089
21      AARTSELAAR     2015         846.570191                9.568364
22      AARTSELAAR     2016         955.191323               12.830730
23      AARTSELAAR     2017         937.840567               -1.816469
24        AFFLIGEM     2010         345.928021                0.000000
25        AFFLIGEM     2011         379.867416                9.811115
26        AFFLIGEM     2012         365.758508               -3.714166
27        AFFLIGEM     2013         390.472345                6.756873
28        AFFLIGEM     2014         213.630255              -45.289274
29        AFFLIGEM     2015         362.630299               69.746696
30        AFFLIGEM     2016         466.280851               28.582982
31        AFFLIGEM     2017         609.738644               30.766392
32  AISEAU-PRESLES     2010         202.224760                0.000000
33  AISEAU-PRESLES     2011         221.121605                9.344477
34  AISEAU-PRESLES     2012         239.538890                8.329030
35  AISEAU-PRESLES     2013         225.664047               -5.792313
36  AISEAU-PRESLES     2014         199.000213              -11.815721
37  AISEAU-PRESLES     2015         263.067455               32.194559
38  AISEAU-PRESLES     2016         214.632766              -18.411509
39  AISEAU-PRESLES     2017         203.303090               -5.278633
40           ALKEN     2010         182.980757                0.000000
41           ALKEN     2011         124.337789              -32.048708
42           ALKEN     2012         154.642079               24.372550
43           ALKEN     2013         119.200549              -22.918425
44           ALKEN     2014         221.172857               85.546844
45           ALKEN     2015         180.365980              -18.450219
46           ALKEN     2016         247.547174               37.247154
47           ALKEN     2017         282.550456               14.140045
48           ALOST     2010         407.869516                0.000000
49           ALOST     2011         515.084632               26.286622
50           ALOST     2012         477.041720               -7.385760
51           ALOST     2013         503.069966                5.456178
52           ALOST     2014         469.519069               -6.669231
53           ALOST     2015         593.658164               26.439628
54           ALOST     2016         606.654397                2.189178
55           ALOST     2017   

In [53]:
pd.set_option('display.max_rows', 100)
df_grouped_houses_totalTransactions = df_grouped_houses_totalTransactions.fillna(0)
df_grouped_houses_totalTransactions.head(100)

CD_REFNIS_FR  CD_YEAR  MS_TOTAL_TRANSACTIONS  pct_change_transactions
0           AALTER     2010                     70                 0.000000
1           AALTER     2011                     69                -1.428571
2           AALTER     2012                     60               -13.043478
3           AALTER     2013                    100                66.666667
4           AALTER     2014                     87               -13.000000
5           AALTER     2015                     57               -34.482759
6           AALTER     2016                     73                28.070175
7           AALTER     2017                     85                16.438356
8         AARSCHOT     2010                    138                 0.000000
9         AARSCHOT     2011                    158                14.492754
10        AARSCHOT     2012                    160                 1.265823
11        AARSCHOT     2013                    168                 5.000000
12        AARSCHOT     2014                    198                17.857143
13        AARSCHOT     2015                    128               -35.353535
14        AARSCHOT     2016                    145                13.281250
15        AARSCHOT     2017                    167                15.172414
16      AARTSELAAR     2010                     90                 0.000000
17      AARTSELAAR     2011                     79               -12.222222
18      AARTSELAAR     2012                     83                 5.063291
19      AARTSELAAR     2013                     85                 2.409639
20      AARTSELAAR     2014                     97                14.117647
21      AARTSELAAR     2015                     76               -21.649485
22      AARTSELAAR     2016                     78                 2.631579
23      AARTSELAAR     2017                    108                38.461538
24        AFFLIGEM     2010                     93                 0.000000
25        AFFLIGEM     2011                     86                -7.526882
26        AFFLIGEM     2012                     98                13.953488
27        AFFLIGEM     2013                     89                -9.183673
28        AFFLIGEM     2014                     85                -4.494382
29        AFFLIGEM     2015                     66               -22.352941
30        AFFLIGEM     2016                     97                46.969697
31        AFFLIGEM     2017                     81               -16.494845
32  AISEAU-PRESLES     2010                     90                 0.000000
33  AISEAU-PRESLES     2011                     76               -15.555556
34  AISEAU-PRESLES     2012                     78                 2.631579
35  AISEAU-PRESLES     2013                     61               -21.794872
36  AISEAU-PRESLES     2014                     54               -11.475410
37  AISEAU-PRESLES     2015                     75                38.888889
38  AISEAU-PRESLES     2016                     75                 0.000000
39  AISEAU-PRESLES     2017                     75                 0.000000
40           ALKEN     2010                     39                 0.000000
41           ALKEN     2011                     38                -2.564103
42           ALKEN     2012                     45                18.421053
43           ALKEN     2013                     36               -20.000000
44           ALKEN     2014                     34                -5.555556
45           ALKEN     2015                     40                17.647059
46           ALKEN     2016                     41                 2.500000
47           ALKEN     2017                     54                31.707317
48           ALOST     2010                    738                 0.000000
49           ALOST     2011                    813                10.162602
50           ALOST     2012                    766                -5.781058
51           ALOST     2013                    736    

In [ ]:
df_grouped_houses_price_per_surface

#### Validation Split

The Validation Split is where you partition your data into a
training set, and a validation set (and sometimes a
test set as well).
You train the model with the training set only. The
validation and/or test sets, are held-back from training and
are used to assess model performance. They provide a
true understanding of how accurate predictions are on
new or unseen data.
An approach called k-fold cross validation can provide
you with an even more robust understanding of model
performance.
Here, the entire dataset is again partitioned into training &
validation sets, and the model is trained and assessed like
before. However, this process is done multiple times with
the training and test sets being rotated to encompass
different sets of observations within the data. You do this
k times - with your final predictive accuracy assessment
being based upon the average of each of the iterations.

In general, validation split is not needed for forecasting problems. Forecasting is the process of predicting future values of a time series based on its past values. In this context, the goal is to accurately predict future values, not to evaluate the performance of a model on a holdout dataset.

### Summary  

In this project we were able to take a dirty dataset and go step by step on a data cleaning and preparation journey.
We introduced CRISP-DM to explain the different phases of a data science project, but we focused on data understanding and data preparation phases.
For Data preparation we used the 8 step model defined by Data Science Infinity which clearly explains the steps that are usually done on data cleaning scenario.

### Further Steps

With the dataframes already cleaned, we could do the following:

1. Create line charts to see the Nr of Transactions per year per municipality.
2. Do the same but use the % change for Transactions and price per square meter, this will allow the users to see where the square meter prices are increasing faster.
3. Others, send your feedback